src/cache/gcs.rs

```diff
@@ -0,0 +1,353 @@
+                        Err(e.into())
+                    }
+                }
+            })
+        }))
+    }
+}
+
+pub struct GCSCredentialProvider {
+    read_only: bool,
+    credentials_path: String,
+    cached_credentials: RefCell<Option<Shared<SFuture<GCSCredential>>>>,
+}
+
+#[derive(Debug, Deserialize)]
+struct ServiceAccountKey {
+    #[serde(rename = "type")]
+    _type: String,
+    project_id: String,
+    private_key_id: String,
+    private_key: String,
+    client_email: String,
+    client_id: String,
+    auth_uri: String,
+    token_uri: String,
+    auth_provider_x509_cert_url: String,
+}
+
+#[derive(Serialize)]
+struct JwtClaims {
+    #[serde(rename = "iss")]
+    issuer: String,
+    scope: String,
+    #[serde(rename = "aud")]
+    audience: String,
+    #[serde(rename = "exp")]
+    expiration: i64,
+    #[serde(rename = "iat")]
+    issued_at: i64,
+}
+
+#[derive(Deserialize)]
+struct TokenMsg {
+    access_token: String,
+    token_type: String,
+    expires_in: usize,
+}
+
+#[derive(Clone)]
+pub struct GCSCredential {
+    token: String,
+    expiration_time: chrono::DateTime<chrono::UTC>,
+}
+
+impl GCSCredentialProvider {
+    pub fn new(read_only: bool, credentials_path: String) -> Self {
+        GCSCredentialProvider {
+            read_only,
+            credentials_path,
+            cached_credentials: RefCell::new(None),
+        }
+    }
+
+    fn auth_request_jwt(&self, expire_at: &chrono::DateTime<chrono::UTC>) -> Result<String> {
+        let metadata = fs::metadata(&self.credentials_path).chain_err(|| {
+            "Couldn't stat GCS credentials file"
+        })?;
+        if !metadata.is_file() {
+            bail!("Couldn't open GCS credentials file.");
+        }
+        let mut file = File::open(&self.credentials_path)?;
+        let mut service_account_json = String::new();
+        file.read_to_string(&mut service_account_json)?;
+        let sa_key: ServiceAccountKey = serde_json::from_str(&service_account_json)?;
+
+        let scope = (if self.read_only {
+            "https://www.googleapis.com/auth/devstorage.readonly"
+        } else {
+            "https://www.googleapis.com/auth/devstorage.read_write"
+        }).to_owned();
+
+        let jwt_claims = JwtClaims {
+            issuer: sa_key.client_email,
+            scope: scope,
+            audience: "https://www.googleapis.com/oauth2/v4/token".to_owned(),
+            expiration: expire_at.timestamp(),
+            issued_at: chrono::UTC::now().timestamp(),
+        };
+
+        let binary_key = openssl::rsa::Rsa::private_key_from_pem(
+            sa_key.private_key.as_bytes()
+        )?.private_key_to_der()?;
+
+        let auth_request_jwt = jwt::encode(
+            &jwt::Header::new(jwt::Algorithm::RS256),
+            &jwt_claims,
+            &binary_key,
+        )?;
+
+        Ok(auth_request_jwt)
+    }
+
+    pub fn credentials(&self, client: &Client<HttpsConnector>) -> SFuture<GCSCredential> {
```

Comment:
I think you should split this function up a bit--it's got a lot going on and it's hard to follow it all.